### DISCUSSION TITLE
Large corrected variables with MOGA

### DISCUSSION THREAD

**Discussion by Val4on** [snl-dakota/dakota#14] [Q&A] [answered]
I'm getting the following message:
23:29:18: quiet- method_MOGA: Invalid variable rep of 3.32079 found for variable var_X Corrected to 2 (value=2.5e-07).       
23:29:18: quiet- method_MOGA: Invalid variable rep of 18.3756 found for variable Var_Y  Corrected to 33 (value=1.8e-06).        
.....

I'm using Moga with a surrogate and offset_uniform as the mutation_type.

**Comment by jadamstephens** [comment, marked as answer]
Hi,

I asked the developer of JEGA, which is the library that provides the MOGA method to Dakota. His name is John Eddy. His response is quoted below.

> .... there is a check in the main algorithm loop that verifies that any values produced by the variational operators are valid and within bounds.  The assumption here is that there may be evaluators that cannot deal with invalid and/or out of bounds variables.  This may especially be true for searches over discrete spaces.
>
> The behavior of the correction is as follows:
> 1. If a value is valid, nothing is done.
> 2. If a value is out of bounds, a new random valid value is substituted.  It used to be the case that the value was corrected to the nearest bounding value.  However, that biased the search toward the bounding values depending on choices of operator so a random value is now used.
> 3. If a value is in bounds but invalid, the nearest valid value is found and substituted (for example, a 1.5 may become a 2 in a discrete space).
>
> The reason you see something like “variable Var_Y Corrected to 33 (value=1.8e-06).” with a corrected number and a corresponding value is because for some spaces, particularly discrete spaces, the algorithm actually operates on indexes into a list.  That is referred to as the variable rep in the innards of JEGA.  The rep can always be decoded into an actual value.  So in this case, the discrete value with index 33 is 1.8e-06.
>
> The offset uniform will likely produce a lot of invalid values over a discrete space.  Many of them will be correctable to the nearest index but those that fall out of bounds will be randomized.  The offset mutators may or may not make sense in a discrete set optimization depending on the locality/relationship of values in the discrete set.  The offset does what it sounds like and tries to produce a value that is statistically close to an existing value.  In a discrete case, that will amount to generating a new index that is statistically close to an existing one.  The resulting values may or may not have anything to do with one another which may defeat the purpose of the operation in the first place.  Finally, if the statistics are tight, then there is a chance of many or most mutation operations being corrected right back to the original value and thus having no effect.  Think of a case where a value is normal 3-sigma within 5% of the existing value relative to a total range of 10.  Then an index, say 3, will most often be mutated to something in the range of 2.5-3.5.  Most of that range will be corrected back to 3.  Only 3.5 may get rounded up to 4.  Values outside 3-sigma on the high side will result in 4 or higher.. low side to 2 or lower.  Something like >99% of the time, no change will occur.

**Reply by Val4on** [reply]
Awesome. This makes a lot of sense. Thank you.
